### ISSUE TITLE
Admission Control Filter (Client-side Throttling)

### ISSUE THREAD

**Issue by tonya11en** [envoyproxy/envoy#9658]
# Admission Control Filter

_**TL; DR:** Probabilistically reject outbound requests based on upstream success-rate._

In scenarios where a single request does not have predictable resource utilization, it can be difficult to configure circuit breaker values. Additionally, circuit breakers are unable to protect against connection thrashing, so we rely on [local ratelimiting](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/local_rate_limit_filter) which suffers from the same problems.

It is possible to protect upstreams more generically by detecting when a significant portion of recent requests to an upstream have been rejected and probabilistically reject requests destined for it before sending it on the wire. Finagle calls this [admission control](https://twitter.github.io/finagle/guide/Clients.html#admission-control) and the Google SRE book calls it [client-side throttling](https://landing.google.com/sre/sre-book/chapters/handling-overload/). Envoy would benefit from a filter that provides this functionality.

The general approach is outlined here, but the idea is that you keep track of the request success-rate over a rolling window and reject outbound requests with a probability of:
![image](https://lh3.googleusercontent.com/S-HeyVOL7ZbrEPaCmKMoJelAKDntPKvNp0j3q7_n_o-At5efdDLhR6Rd_N9Oj9HfMLLjtYu-XBQK34UJrDWPuXOA4eSU0kI-iOP0zg=s0)

We would be able to make use of a configurable measure of success, similar to the retry policy's [retry_on](https://www.envoyproxy.io/docs/envoy/v1.7.0/api-v1/route_config/route#config-http-conn-man-route-table-route-retry) parameter.

**Comment by mattklein123** [comment]
Calling this done with the latest PR that @tonya11en landed. We can open new issues for smaller items as needed. Thank you @tonya11en!
